app/actions/views/permalink.ts

```diff
@@ -2,44 +2,113 @@
-                    onClose: () => {
-                        dispatch(closePermalink());
-                    },
-                };
-
-                const options = {
-                    layout: {
-                        componentBackgroundColor: changeOpacity('#000', 0.2),
-                    },
-                };
-
-                showingPermalink = true;
-                showModalOverCurrentContext(screen, passProps, options);
+
+            // Warn before joining private channels through permalink
+            const state = getState();
+            const currentUserId = getCurrentUserId(state);
+            const roles = currentUserId ? getCurrentUserRoles(state) : '';
+            if (isSystemAdmin(roles)) {
+                const postData = await dispatch(getPost(postId));
+                if (postData.error) {
+                    return {error: postData.error};
+                }
+                const channelId = postData.data?.channel_id;
+                if (channelId) {
+                    const {data: isMember} = isChannelMember(state, channelId);
+                    if (!isMember) {
+                        const {data: channelData, error: channelError} = await dispatch(getChannel(channelId));
+                        if (channelError) {
+                            return {error: channelError};
+                        }
+                        if (channelData.type === General.PRIVATE_CHANNEL) {
+                            Alert.alert(
+                                intl.formatMessage({
+                                    id: 'permalink.show_dialog_warn.title',
+                                    defaultMessage: 'Private Channel',
+                                }),
+                                intl.formatMessage({
+                                    id: 'permalink.show_dialog_warn.description',
+                                    defaultMessage: 'You must join "{channel}" to view this post.',
+                                }, {
+                                    channel: channelData.name,
+                                }),
+                                [
+                                    {
+                                        text: intl.formatMessage({
+                                            id: 'permalink.show_dialog_warn.cancel',
+                                            defaultMessage: 'Cancel',
+                                        }),
+                                    },
+                                    {
+                                        text: intl.formatMessage({
+                                            id: 'permalink.show_dialog_warn.join',
+                                            defaultMessage: 'Join',
+                                        }),
+                                        onPress: async () => {
+                                            await dispatch(joinChannel(currentUserId, teamName, channelId, channelData.name));
```

Comment:
What happens if this fails? Can a sustem admin join a private channel in any version of MM?